### ISSUE TITLE
Work with "separator-based" data

### ISSUE THREAD

**Issue by notowen333** [privacy-tech-lab/privacy-pioneer#357]
## Overview 

There is a class of "separator-based" user data that has a similar shape and should be generalized in the codebase and UI. Examples of "separator-based" data include:

- Email address `asdf@gmail.com`
- Credit Card Number `4464 6846 4354 3564` or `4464-6846-4354-3564` etc.
- Social Security Number `111-02-1211` or `111 02 1211` etc.
- IP Address `123.22.11.1`
- Phone Number `123-456-7890`

### Regular Expressions

When users give us this kind of data, we want to automatically create REs which make the separators (`"@"`, `"."`, `"-"`, `":"`, etc) optional. We also want to automatically deal with escape sequences for "`.`" because regular expressions use this character as a wildcard.

So we want `"asdf@gmail.com"` -> `/asdf@?gmail\.?com/i`

Our motivation is to try to catch as many syntaxes of these user data as possible. 
#### Email Notes

1. For emails, we want to make sure that we are making our REs case insensitive ("i" flag in the above example)
2. Maybe also we want to skip endings like ".com", ".org", or ".edu" as well.

### UI

In the watchlist, we may want to create a category called something like "separator data"  which performs this routine. We can give examples like credit card and ssn to explain the concept. Alternatively, this could be a routine that happens automatically when a "general" keyword is added.

### Implementation 

We already have categories and routines for IP Address and Phone Number, so maybe nothing needs to be done here. Email has a category but no routine, so definitely changes are needed there. The generalized "separator data" category does not currently exist, so this needs to be implemented. 

Creating the actual regular expressions shouldn't be too challenging, but the overall task requires working with many different areas of the codebase.

**Comment by SebastianZimmeck** [comment]
If this is fairly quick to do, i.e., a new regex and some UI database changes, I'd say yes. Otherwise, we have to think a bit where to spend our resources. Let's discuss on Friday.

**Comment by SebastianZimmeck** [comment]
Excellent, thank you, @danielgoldelman! 

> For email addresses, we already look for the most used obfuscation methods of emails. 

What obfuscation are we actually looking at?

> The most pressing issue not mentioned above is state regexes, as we currently are only looking at US states for the user data. This needs to be changed, and will require a pretty big overhaul.

I do not find states particularly exciting. Also, if a site already receives a user's email address, it will already have a much more precise location (assuming the user is not using a VPN).

**Comment by danielgoldelman** [comment]
> What obfuscation are we actually looking at?

We look at various conversions of an email address from the normal string (eg "abc@gmail.com") to a few other versions: the hex value SHA-246 encoded version, and the base64 SHA-256 encoded version. These versions are normally used instead of the normal string because they remove the "@" symbol and make it less readable for an end user. Also the encoding/decoding is written into javascript, so it's very simple for them to use.

> states comments

Removing the states type is certainly something we should talk about at our next meeting. They work really well in the United States because it is so simple to map zip code to states, but generalizing the process is going to be quite annoying. Lets discuss later.

**Comment by SebastianZimmeck** [comment]
> We look at various conversions of an email address

OK, great! Could you add this to the readme. I do not think that we describe this anywhere yet. So, basically we are looking at some cryptographic hashing (SHA-256), some standard formatting (Base64), and their combination.

(I am not sure what SHA-246 is; is that a typo?) 

> Removing the states type is certainly something we should talk about at our next meeting.

I am not necessarily for removing it (also not against it; I am neutral on ZIP codes :) ). Just, if it is a lot of work, we should discuss where to spend our resources.

**Comment by danielgoldelman** [comment]
^ General keyword type regex changes implemented as described above. 

Yes, SHA-246 was a typo. Fixed above.

**Comment by SebastianZimmeck** [comment]
Thanks for clarifying, @danielgoldelman! 

If you could could also add to the readme how we go about the email address.

**Comment by SebastianZimmeck** [comment]
Per our discussion today, @danielgoldelman will look into whether the particular encoding and encryption is actually used in practice. We believe we implemented it this way because of some guidance we found on the Internet by an ad network or something similar. @danielgoldelman will try to dig up our conversation at the time and learn more about whether these encodings and encryptions are used in practice.
